src/util/type.cpp

```diff
@@ -34,3 +36,72 @@ bool is_number(const typet &type)
          id==ID_floatbv ||
          id==ID_fixedbv;
 }
+
+/// Identify if a given type is constant itself or
+/// contains constant components. Examples include:
+///  - const int a;
+///  - struct contains_constant_pointer {  int x; int * const p; };
+///  - const int b[3];
+/// \param type The type we want to query constness of.
+/// \param ns The namespace, needed for resolution of symbols.
+/// \return Whether passed in type is const or not.
+bool is_constant_or_has_constant_components(
+  const typet &type, const namespacet &ns)
+{
+  // Helper function to avoid the code duplication
+  // in the branches below.
+  const auto has_constant_components =
+  [](const typet &subtype) -> bool
+  {
+    if(subtype.id() == ID_struct || subtype.id() == ID_union)
+    {
+      const auto &struct_union_type = to_struct_union_type(subtype);
+      for(const auto &component : struct_union_type.components())
+      {
+        if(component.type().id() == ID_pointer)
+          return component.type().get_bool(ID_C_constant);
```

Comment:
Why do you return `false` early in this case? There could be a constant member?

Reply:
Ah, good catch!